File: domain/publish.py
"""Publish job state machine and target results."""

from dataclasses import dataclass, replace
from datetime import datetime
from enum import StrEnum
from uuid import UUID


class PublishJobStatus(StrEnum):
    """Durable job states from draft through independent verification."""

    DRAFT = "draft"
    PREFLIGHT = "preflight"
    AWAITING_CONFIRMATION = "awaiting_confirmation"
    PUBLISHING = "publishing"
    SWITCHING = "switching"
    VERIFYING = "verifying"
    COMPLETED = "completed"
    PARTIAL_FAILURE = "partial_failure"
    FAILED = "failed"


class TargetStatus(StrEnum):
    """Outcome of one station target in a publish run."""

    SIMULATED = "simulated"
    VERIFIED = "verified"
    ERROR = "error"
    RECOVERY_REQUIRED = "recovery_required"


ALLOWED_TRANSITIONS: dict[PublishJobStatus, frozenset[PublishJobStatus]] = {
    PublishJobStatus.DRAFT: frozenset({PublishJobStatus.PREFLIGHT}),
    PublishJobStatus.PREFLIGHT: frozenset({PublishJobStatus.AWAITING_CONFIRMATION}),
    PublishJobStatus.AWAITING_CONFIRMATION: frozenset({PublishJobStatus.PUBLISHING}),
    PublishJobStatus.PUBLISHING: frozenset(
        {PublishJobStatus.SWITCHING, PublishJobStatus.COMPLETED, PublishJobStatus.FAILED}
    ),
    PublishJobStatus.SWITCHING: frozenset(
        {PublishJobStatus.VERIFYING, PublishJobStatus.PARTIAL_FAILURE, PublishJobStatus.FAILED}
    ),
    PublishJobStatus.VERIFYING: frozenset(
        {PublishJobStatus.COMPLETED, PublishJobStatus.PARTIAL_FAILURE, PublishJobStatus.FAILED}
    ),
    PublishJobStatus.COMPLETED: frozenset(),
    PublishJobStatus.PARTIAL_FAILURE: frozenset(),
    PublishJobStatus.FAILED: frozenset(),
}


@dataclass(frozen=True, slots=True)
class PublishJob:
    """Immutable job state with explicit legal transitions."""

    id: UUID
    idempotency_key: str
    correlation_id: UUID
    label: str
    source_dataset: str
    dry_run: bool = True
    allow_hot_switch: bool = False
    status: PublishJobStatus = PublishJobStatus.DRAFT
    description: str | None = None
    status_reason: str | None = None
    operator_id: UUID | None = None
    client_confirmation: bool | None = None
    client_confirmation_at: datetime | None = None

    def transition(self, target: PublishJobStatus) -> "PublishJob":
        if target not in ALLOWED_TRANSITIONS[self.status]:
            raise ValueError(f"illegal publish transition: {self.status} -> {target}")
        return replace(self, status=target)


@dataclass(frozen=True, slots=True)
class PublishTarget:
    """Materialized station selection and per-target publish state."""

    id: UUID
    job_id: UUID
    station_id: UUID
    selected: bool = True
    preflight_status: str | None = None
    preflight_result: dict[str, object] | None = None
    old_version_id: UUID | None = None
    new_version_id: UUID | None = None
    old_mapping: dict[str, object] | None = None
    new_mapping: dict[str, object] | None = None
    switch_status: str | None = None
    verify_status: str | None = None
    error_code: str | None = None
    error_message: str | None = None
    progress_percent: int = 0

    def __post_init__(self) -> None:
        if not 0 <= self.progress_percent <= 100:
            raise ValueError("progress_percent must be between 0 and 100")


@dataclass(frozen=True, slots=True)
class PublishTargetResult:
    """Safe result for one station; old mapping is retained in the result."""

    station_id: UUID
    status: TargetStatus
    old_mapping: str | None
    new_mapping: str | None = None
    error_code: str | None = None
    error_message: str | None = None

File: frontend/src/domain/publish.ts
export type CheckStatus = "pass" | "block" | "unknown" | "warning";

export interface PreflightCheck {
  status: CheckStatus;
  code: string;
  message: string;
  observed_at: string;
  source_snapshot_id: string | null;
}

export interface PreflightReport {
  station_id: string;
  status: CheckStatus;
  can_publish: boolean;
  evaluated_at: string;
  checks: PreflightCheck[];
}

export type PublishJobStatus =
  | "draft"
  | "preflight"
  | "awaiting_confirmation"
  | "publishing"
  | "switching"
  | "verifying"
  | "completed"
  | "partial_failure"
  | "failed";

export interface PublishJobDraft {
  id: string;
  idempotency_key: string;
  correlation_id: string;
  label: string;
  source_dataset: string;
  status: PublishJobStatus;
  dry_run: boolean;
  allow_hot_switch: boolean;
  station_ids: string[];
}

export interface PublishGate {
  status: "ready" | "blocked";
  can_advance: boolean;
  selected_station_ids: string[];
  reasons: string[];
}

export interface PublishPrepareResponse {
  job_id: string;
  status: PublishJobStatus;
  client_confirmation: boolean | null;
  gate: PublishGate;
  admin_report: PreflightReport | null;
  client_reports: PreflightReport[];
}

export interface PublishDispatchResponse {
  job_id: string;
  status: PublishJobStatus;
  accepted: boolean;
}

export interface PublishTargetReadModel {
  station_id: string;
  selected: boolean;
  preflight_status: string | null;
  switch_status: string | null;
  verify_status: string | null;
  error_code: string | null;
  error_message: string | null;
  progress_percent: number;
}

export interface PublishJobReadModel {
  id: string;
  idempotency_key: string;
  correlation_id: string;
  label: string;
  source_dataset: string;
  description: string | null;
  status: PublishJobStatus;
  dry_run: boolean;
  allow_hot_switch: boolean;
  targets: PublishTargetReadModel[];
}

export const checkStatusLabel: Record<CheckStatus, string> = {
  pass: "PASS",
  block: "BLOCK",
  unknown: "UNKNOWN",
  warning: "WARNING",
};

export const checkStatusDescription: Record<CheckStatus, string> = {
  pass: "Проверка пройдена.",
  block: "Проверка блокирует публикацию.",
  unknown: "Нельзя подтвердить безопасность по текущим данным.",
  warning: "Есть предупреждение; решение остаётся за серверной политикой.",
};

export const publishStatusLabel: Record<PublishJobStatus, string> = {
  draft: "Черновик",
  preflight: "Preflight",
  awaiting_confirmation: "Ожидает подтверждения",
  publishing: "Принято в публикацию",
  switching: "Переключение",
  verifying: "Проверка",
  completed: "Завершено",
  partial_failure: "Частичный сбой",
  failed: "Ошибка",
};
